scipy/sparse/linalg/_special_sparse_arrays.py

```diff
@@ -511,3 +513,391 @@ def _adjoint(self):
+        shape = (n, n)
+        super().__init__(dtype, shape)
+
+        k = np.arange(1, n + 1)
+        e = np.sort(16. * np.power(np.cos(0.5 * k * np.pi / (n + 1)), 4))
+        self._eigenvalues = e
+
+    @property
+    def eigenvalues(self):
+        return self._eigenvalues
+
+    @eigenvalues.setter
+    def eigenvalues(self, value):
+        raise AttributeError('"eigenvalues" attribute is read-only and cannot '
+                             'be set.')
+
+    def tosparse(self):
+        from scipy.sparse import spdiags
+        d0 = np.r_[5, 6 * np.ones(self.n - 2, dtype=self.dtype), 5]
+        d1 = -4 * np.ones(self.n, dtype=self.dtype)
+        d2 = np.ones(self.n, dtype=self.dtype)
+        return spdiags([d2, d1, d0, d1, d2], [-2, -1, 0, 1, 2], self.n, self.n)
+
+    def tobanded(self):
+        d0 = np.r_[5, 6 * np.ones(self.n - 2, dtype=self.dtype), 5]
+        d1 = -4 * np.ones(self.n, dtype=self.dtype)
+        d2 = np.ones(self.n, dtype=self.dtype)
+        return np.array([d2, d1, d0])
+
+    def toarray(self):
+        d0 = np.r_[5, 6 * np.ones(self.n - 2, dtype=self.dtype), 5]
+        d1 = -4 * np.ones(self.n - 1, dtype=self.dtype)
+        d2 = np.ones(self.n - 2, dtype=self.dtype)
+        a = np.diag(d0)
+        a += np.diag(d1, 1) + np.diag(d1, -1)
+        a += np.diag(d2, 2) + np.diag(d2, -2)
+        return a
+    
+    def _matvec(self, x):
+        n = self.n
+        assert n == x.shape[0]
+        x = x.reshape(n, -1)
+        sx = np.zeros_like(x)
+        sx[0,:] = 5 * x[0,:] - 4 * x[1,:] + x[2,:]
+        sx[-1,:] = 5 * x[-1,:] - 4 * x[-2,:] + x[-3,:]
+        sx[1:-1,:] = (6 * x[1:-1,:] - 4 * (x[:-2,:] + x[2:,:])
+                      + np.pad(x[:-3,:], ((1,0),(0,0)))
+                      + np.pad(x[3:,:], ((0,1),(0,0))))
+        return sx
+
+    def _matmat(self, x):
+        return self._matvec(x)
+
+    def _adjoint(self):
+        return self
+
+    def _transpose(self):
+        return self
+
+
+class MikotaM(LinearOperator):
+    """
+    Construct a mass matrix in various formats of Mikota pair.
+
+    The mass matrix `M` is square real diagonal
+    positive definite with entries that are reciprocal to integers.
+
+    Parameters
+    ----------
+    shape : tuple of int
+        The shape of the matrix.
+    dtype : dtype
+        Numerical type of the array. Default is ``np.float64``.
+
+    Methods
+    -------
+    toarray()
+        Construct a dense array from Laplacian data
+    tosparse()
+        Construct a sparse array from Laplacian data
+    tobanded()
+        The format for banded symmetric matrices,
+        i.e., (1, n) ndarray with the main diagonal.
+    """
+    def __init__(self, shape, dtype):
+        self.shape = shape
+        self.dtype = dtype
+        super().__init__(dtype, shape)
+
+    def tosparse(self):
+        from scipy.sparse import diags
+        n = self.shape[0]
+        aranp1 = np.arange(1, n + 1, dtype=self.dtype)
+        aranp1_inv = 1. / aranp1
```

Comment:
Aside from unclear, unexplained naming, it seems like we might want to pre-calculate the quantity `aranp1_inv` and store it in the class. Not so much for performance, but because it would deduplicate the 4 times this is calculated in the methods.

Reply:
`aranp1 =  arange_plus1` 
I do not know how to "pre-calculate the quantity aranp1_inv and store it in the class". Could you please suggest specific changes?

Reply:
```suggestion
```

Reply:
@h-vetinari thanks for explaining your suggestion with the specific proposal for changes! However, as I now understand your suggestion, I am unsure if it is a proper approach for the custom `LinearOperator` classes in this PR, although the code is shorter/nicer of course.

The point of the `LinearOperator` infrastructure is to be matrix-free, i.e., avoid constructing matrices unless explicitly required. Moreover, the standard `LinearOperator` class does not even have the `toarray` or `tosparse` or 'banded' parameters to generate the matrices; see https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.LinearOperator.html
There are added here for user convenience but are kind-of optional. Precomputing matrices in init makes the code shorter in these optional parts but appears to be a waste for the main matrix-free API which is the `matvec`. Let's ask @rkern to express his opinion on such changes.

Reply:
I also don't see the issue with saving another vector (you already have `self._eigenvalues` saved, which is the same size), but if storage footprint is an issue (we're far from a full matrix, but OK), then you can make `self._diag()` a member function that computes the thing on demand. It would lead to the same amount of code deduplication. Same for `self._bands()` in Sakurai.

Reply:
All diagonals can be grabbed from the already available `tobanded` method without repetition so no need to write a new `self._diag()`  member function. I am making the changes now.

Reply:
Doing this with `tobanded` is not the same (talking about Sakurai here, as Mikota already uses `self._diag` now), because it returns an array, whereas my proposal is still a list. Crucially, we can very easily extract the components from the latter, but not the former:
```
d2, d1, d0 = self._bands()  # works without extra ceremony
```